### PR TITLE
test battery: Remove pyc files before running

### PR DESCRIPTION
**Pull request by arjclark**
If errant .pyc files are around it can result in code working when it shouldn't be - hence the addition of a new test battery test in #1439.

However, we really need to know whether or not our pyc files line up before running any other tests. Otherwise you could spend 15+ minutes running the test battery before finding you need to remove the file and re-run.

~~This moves the test from #1439 to the cylc-test-battery script itself so it performs the test and aborts if an errant .pyc file is present.~~ (edited)

This change clears out all the pyc files under $CYLC_DIR so we're running with only the code we're working with.  

@benfitzpatrick - please review.

### CODE REVIEW

File: bin/cylc-test-battery [hunk 1]
```diff
@@ -80,6 +80,16 @@ if [[ "$PWD" != "$CYLC_DIR" ]]; then
...
     cd "$CYLC_DIR"
 fi
+
+echo "[INFO] Checking .pyc files match .py files..."
+py_files=$(find "$CYLC_DIR" -name "*.pyc" -type f | sed "s/pyc$/py/g")
+ls $py_files 1> /dev/null
+if [ $? -gt 0 ]; then
+    echo "[ERROR] Aborting test battery, .pyc files do not match .py files"
+else
+    echo "[INFO] ...ok"
```
Comment (benfitzpatrick) [inline bin/cylc-test-battery]: I would just put 'pyc files ok' here and avoid the first INFO line

File: bin/cylc-test-battery [hunk 2]
```diff
@@ -80,6 +80,16 @@ if [[ "$PWD" != "$CYLC_DIR" ]]; then
     echo "[INFO] cd \"$CYLC_DIR\""
     cd "$CYLC_DIR"
 fi
+
+echo "[INFO] Checking .pyc files match .py files..."
+py_files=$(find "$CYLC_DIR" -name "*.pyc" -type f | sed "s/pyc$/py/g")
+ls $py_files 1> /dev/null
+if [ $? -gt 0 ]; then
```
Comment (benfitzpatrick) [inline bin/cylc-test-battery]: Probably better to use 'ls' in the if - `if ls $py_files 1>/dev/null; then`

**Comment by hjoliver** [comment]
Can we not just do ```find $CYLC_DIR -name '*.pyc' | xargs rm``` in ```bin/cylc-test-battery```, before running the tests?

**Comment by arjclark** [comment]
@benfitzpatrick - any objections to @hjoliver's suggestion? I think it's preferable but there's a niggle at the back of my mind that this might not be ideal.

**Comment by matthewrmshin** [comment]
As a aside... The `*.pyc` files are really more trouble than they are worth.
* They get left behind.
* They complicates installation.
* They work only if you have a single Python 2.X at your site.
* They don't really speed up your program by much. (I have done some tests.)

If you have `*.pyc` files for a Python 2.X (or have no `*.pyc` files at all) in your site installation of cylc, and 2 user processes attempt to run cylc with an alternate Python 2.X, both will attempt to write `*.pyc` files which will fail because they don't have write permission. On some file systems, we have also observed a massive slow down when this happens.

Perhaps we should have `export PYTHONDONTWRITEBYTECODE=true` in our environment.

**Comment by hjoliver** [comment]
Apparently pyc files load faster at start-up.  But I've never tested that, and I agree the disadvantages are annoying.

**Comment by matthewrmshin** [comment]
> Apparently pyc files load faster at start-up. But I've never tested that, and I agree the disadvantages are annoying.

We have tested this a few months ago in local file systems, file systems mounted via NFS, etc. The speed up `*.pyc` files offer was pretty much non-existent.

**Comment by benfitzpatrick** [comment]
> @benfitzpatrick - any objections to @hjoliver's suggestion?

I think this is OK as long as we are careful

**Comment by arjclark** [comment]
@benfitzpatrick - Have changed to using @hjoliver's suggested method instead and updated title and description accordingly.

File: bin/cylc-test-battery [hunk 3]
```diff
@@ -80,6 +80,10 @@ if [[ "$PWD" != "$CYLC_DIR" ]]; then
     echo "[INFO] cd \"$CYLC_DIR\""
     cd "$CYLC_DIR"
 fi
+
+# Remove pyc files to ensure we are running the current code.
+find $CYLC_DIR -name '*.pyc' | xargs rm
```
Comment (benfitzpatrick) [inline bin/cylc-test-battery]: We need a test for the existence of $CYLC_DIR

Comment (hjoliver) [inline bin/cylc-test-battery]: Or ```set -u``` would do.

File: bin/cylc-test-battery [hunk 4]
```diff
@@ -16,6 +16,8 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+set -u
```
Comment (arjclark) [inline bin/cylc-test-battery]: set -u added here

**Comment by arjclark** [comment]
@benfitzpatrick - added in set -u to script.

File: bin/cylc-test-battery (hunk 3 above)
Comment (arjclark) [inline bin/cylc-test-battery]: Added below.

**Pull request merged**